pylint/extensions/typing.py

```diff
@@ -219,6 +224,66 @@ def visit_attribute(self, node: nodes.Attribute) -> None:
         if self._should_check_callable and node.attrname == "Callable":
             self._check_broken_callable(node)
 
+    @only_required_for_messages("redundant-union-assign-typehint")
+    def visit_annassign(self, node: nodes.AnnAssign) -> None:
+        type_annotations = list(node.annotation.get_children())
+        if not type_annotations:
+            return
+        if (
+            isinstance(type_annotations[0], nodes.Name)
+            and type_annotations[0].name in UNION_NAMES
+        ):
+            if isinstance(type_annotations[-1], nodes.Tuple):
+                types = list(type_annotations[-1].get_children())
+            elif type_annotations[0].name == "Optional":
+                if (
+                    isinstance(type_annotations[-1], nodes.Const)
+                    and type_annotations[-1].pytype() == "builtins.NoneType"
+                ):
+                    self.add_message(
+                        "redundant-union-assign-typehint",
+                        node=node,
```

Comment:
The error-node shouldn't be the `AnnAssign`. We can be more specific. For `Optional`, it should probably be `node.annotation`.

For Union, we could even highlight the exact duplicate item (node) 🤔